File: rust/src/filter.rs
//! Message filtering. Both `exclude_*` (negative) and `include_*`
//! (positive) filters are supported. A message passes if:
//!   - it matches no `exclude_*` set, AND
//!   - every active `include_*` set contains its value.
//!
//! Inactive sets (empty) are ignored on both sides.

use crate::models::{Bus, MieMessage};

#[derive(Debug, Clone, Default)]
pub struct FilterConfig {
    pub exclude_types: Vec<u8>,
    pub exclude_rts: Vec<u8>,
    pub exclude_buses: Vec<Bus>,
    pub exclude_subaddresses: Vec<u8>,

    pub include_types: Vec<u8>,
    pub include_rts: Vec<u8>,
    pub include_buses: Vec<Bus>,
    pub include_subaddresses: Vec<u8>,
}

/// The four fields every filter looks at, read off a message once.
///
/// Extracted so the exclude and include halves share one decode rather than
/// each pulling the Command Word apart again.
struct FilterFields {
    message_type: u8,
    bus: Bus,
    rt: Option<u8>,
    subaddress: Option<u8>,
}

impl FilterFields {
    fn of(msg: &MieMessage) -> Self {
        let (rt, subaddress) = msg
            .command_word
            .map_or((None, None), |cw| (Some(cw.rt), Some(cw.subaddress)));
        Self {
            message_type: msg.type_word.message_type,
            bus: msg.type_word.bus,
            rt,
            subaddress,
        }
    }
}

impl FilterConfig {
    pub fn is_active(&self) -> bool {
        !self.exclude_types.is_empty()
            || !self.exclude_rts.is_empty()
            || !self.exclude_buses.is_empty()
            || !self.exclude_subaddresses.is_empty()
            || !self.include_types.is_empty()
            || !self.include_rts.is_empty()
            || !self.include_buses.is_empty()
            || !self.include_subaddresses.is_empty()
    }

    /// True if `msg` should be dropped from output.
    ///
    /// The two halves are genuinely different rules, which is why they are
    /// separate functions rather than one run of guards: an *exclude* set drops
    /// a record that IS listed, while an *include* set drops a record that is
    /// not — and an include set only means anything when it is non-empty.
    pub fn should_exclude(&self, msg: &MieMessage) -> bool {
        let field = FilterFields::of(msg);
        self.excluded_by_negative(&field) || self.excluded_by_positive(&field)
    }

    /// Dropped because the record IS in an exclusion set.
    fn excluded_by_negative(&self, field: &FilterFields) -> bool {
        if self.exclude_types.contains(&field.message_type)
            || self.exclude_buses.contains(&field.bus)
        {
            return true;
        }
        if let Some(rt) = field.rt
            && self.exclude_rts.contains(&rt)
        {
            return true;
        }
        if let Some(sa) = field.subaddress
            && self.exclude_subaddresses.contains(&sa)
        {
            return true;
        }
        false
    }

    /// Dropped because an inclusion set is active and the record is not in it.
    ///
    /// A record with no Command Word — SPURIOUS_DATA — has no RT or subaddress
    /// to match, so an active RT or subaddress include set drops it. That is
    /// deliberate: "keep only RT 5" cannot meaningfully keep a record that
    /// names no RT.
    fn excluded_by_positive(&self, field: &FilterFields) -> bool {
        if !self.include_types.is_empty() && !self.include_types.contains(&field.message_type) {
            return true;
        }
        if !self.include_buses.is_empty() && !self.include_buses.contains(&field.bus) {
            return true;
        }
        if !self.include_rts.is_empty()
            && !field.rt.is_some_and(|rt| self.include_rts.contains(&rt))
        {
            return true;
        }
        if !self.include_subaddresses.is_empty()
            && !field
                .subaddress
                .is_some_and(|sa| self.include_subaddresses.contains(&sa))
        {
            return true;
        }
        false
    }
}

/// Iterator adapter wrapping any `Iterator<Item = MieResult<MieMessage>>`.
///
/// Reports the same diagnostics as the Python `apply_filters` generator: the
/// active filter sets once at construction (INFO), each dropped record (DEBUG),
/// and a passed/excluded tally when the stream is finished (INFO). The tally is
/// emitted from `Drop` rather than at end-of-iteration so it still appears when
/// a consumer stops early — a broken pipe, `| head` — where an end-of-stream
/// hook would never run.
pub struct Filtered<I> {
    inner: I,
    filters: FilterConfig,
    passed: u64,
    excluded: u64,
}

impl<I, E> Iterator for Filtered<I>
where
    I: Iterator<Item = Result<MieMessage, E>>,
{
    type Item = Result<MieMessage, E>;
    fn next(&mut self) -> Option<Self::Item> {
        loop {
            match self.inner.next()? {
                Err(e) => return Some(Err(e)),
                Ok(msg) => {
                    if !self.filters.should_exclude(&msg) {
                        self.passed += 1;
                        return Some(Ok(msg));
                    }
                    self.excluded += 1;
                    log_filtered_out(&msg);
                }
            }
        }
    }
}

impl<I> Drop for Filtered<I> {
    fn drop(&mut self) {
        if self.filters.is_active() {
            crate::log_info!(
                "Filter results: {} passed, {} excluded",
                self.passed,
                self.excluded
            );
        }
    }
}

/// One-time INFO summary of the configured filter sets. `none` for an inactive
/// (empty) set, matching the Python wording.
fn log_active_filters(f: &FilterConfig) {
    // Sorted so the line is stable regardless of the order values were parsed
    // in, and rendered identically on both implementations (Python holds these
    // as sets, whose repr order is not guaranteed).
    fn show<T: std::fmt::Display + Ord + Copy>(v: &[T]) -> String {
        if v.is_empty() {
            return "none".to_string();
        }
        let mut sorted: Vec<T> = v.to_vec();
        sorted.sort_unstable();
        let items: Vec<String> = sorted.iter().map(ToString::to_string).collect();
        format!("[{}]", items.join(", "))
    }
    crate::log_info!(
        "Filtering active: exclude_types={} exclude_rts={} exclude_buses={} \
         exclude_subaddresses={} include_types={} include_rts={} include_buses={} \
         include_subaddresses={}",
        show(&f.exclude_types),
        show(&f.exclude_rts),
        show(&f.exclude_buses),
        show(&f.exclude_subaddresses),
        show(&f.include_types),
        show(&f.include_rts),
        show(&f.include_buses),
        show(&f.include_subaddresses),
    );
}

/// DEBUG line for a message dropped by the filters.
fn log_filtered_out(msg: &MieMessage) {
    let (rt, sa) = match msg.command_word {
        Some(cw) => (cw.rt.to_string(), cw.subaddress.to_string()),
        None => ("-".to_string(), "-".to_string()),
    };
    crate::log_debug!(
        "Filtered out: offset=0x{:X} type=0x{:02X} RT{} SA{} Bus {}",
        msg.file_offset,
        msg.type_word.message_type,
        rt,
        sa,
        msg.type_word.bus
    );
}

/// Extension trait: `iter.filter_messages(cfg)`.
pub trait FilterIterExt: Sized {
    fn filter_messages(self, filters: FilterConfig) -> Filtered<Self>;
}

impl<I, E> FilterIterExt for I
where
    I: Iterator<Item = Result<MieMessage, E>>,
{
    fn filter_messages(self, filters: FilterConfig) -> Filtered<Self> {
        if filters.is_active() {
            log_active_filters(&filters);
        } else {
            crate::log_debug!("No filters active, passing all messages through");
        }
        Filtered {
            inner: self,
            filters,
            passed: 0,
            excluded: 0,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::models::*;

    fn msg(rt: u8, sa: u8, bus: Bus, mt: u8) -> MieMessage {
        MieMessage {
            timestamp: Timestamp::Standard(StandardTimestamp {
                raw_value: 0,
                upper_word: 0,
                lower_word: 0,
            }),
            type_word: TypeWord {
                message_type: mt,
                bus,
                word_count: 5,
                error: false,
                raw: 0,
            },
            message_format: MessageFormat::Receive,
            command_word: Some(CommandWord {
                rt,
                direction: Direction::Receive,
                subaddress: sa,
                data_word_count: 1,
                raw: 0,
            }),
            command_word_2: None,
            status_word: None,
            status_word_2: None,
            data_words: DataWords::new(),
            error_word: None,
            delta: Some(0.0),
            file_offset: 0,
            mux: None,
        }
    }

    /// Requirements: L2-FLT-001
    #[test]
    fn empty_config_is_inactive() {
        let cfg = FilterConfig::default();
        assert!(!cfg.is_active());
        assert!(!cfg.should_exclude(&msg(1, 1, Bus::A, 0x02)));
    }

    /// Requirements: L2-CFG-006
    #[test]
    fn exclude_by_rt() {
        let cfg = FilterConfig {
            exclude_rts: vec![31],
            ..Default::default()
        };
        assert!(cfg.should_exclude(&msg(31, 0, Bus::A, 0x02)));
        assert!(!cfg.should_exclude(&msg(15, 0, Bus::A, 0x02)));
    }

    /// Requirements: L2-CFG-006
    #[test]
    fn exclude_by_type_and_bus() {
        let cfg = FilterConfig {
            exclude_types: vec![0x20],
            exclude_buses: vec![Bus::B],
            ..Default::default()
        };
        assert!(cfg.should_exclude(&msg(1, 1, Bus::A, 0x20)));
        assert!(cfg.should_exclude(&msg(1, 1, Bus::B, 0x02)));
        assert!(!cfg.should_exclude(&msg(1, 1, Bus::A, 0x02)));
    }

    /// Requirements: L3-RS-010
    #[test]
    fn include_filters_drop_non_matches() {
        let cfg = FilterConfig {
            include_rts: vec![15],
            ..Default::default()
        };
        assert!(!cfg.should_exclude(&msg(15, 0, Bus::A, 0x02)));
        assert!(cfg.should_exclude(&msg(14, 0, Bus::A, 0x02)));
    }

    /// Requirements: L2-FLT-001
    #[test]
    fn iterator_adapter() {
        let msgs: Vec<Result<MieMessage, ()>> = vec![
            Ok(msg(15, 1, Bus::A, 0x02)),
            Ok(msg(31, 1, Bus::A, 0x02)),
            Ok(msg(0, 1, Bus::B, 0x02)),
        ];
        let cfg = FilterConfig {
            exclude_rts: vec![31],
            ..Default::default()
        };
        let filtered: Vec<_> = msgs
            .into_iter()
            .filter_messages(cfg)
            .collect::<Result<_, _>>()
            .unwrap();
        assert_eq!(filtered.len(), 2);
        assert_eq!(filtered[0].rt(), Some(15));
        assert_eq!(filtered[1].rt(), Some(0));
    }
}

File: cpp/src/filter.cpp
// SPDX-License-Identifier: Apache-2.0

#define MIE_LOG_MODULE "mie_decoder::filter"

#include "mie/filter.hpp"

#include <algorithm>

#include "mie/log.hpp"
#include "mie/text.hpp"

namespace mie {

namespace {

template <typename T>
bool contains(const std::vector<T>& values, T wanted) {
    return std::find(values.begin(), values.end(), wanted) != values.end();
}

/// A set rendered for the one-time INFO line: `none`, or `[a, b, c]`.
///
/// SORTED, and that is not cosmetic. Python holds these as sets, whose
/// iteration order is not guaranteed; sorting is what makes the three
/// implementations emit the same line for the same configuration rather than
/// the same line most of the time.
template <typename T>
std::string show(const std::vector<T>& values) {
    if (values.empty()) {
        return "none";
    }
    std::vector<T> sorted(values);
    std::sort(sorted.begin(), sorted.end());
    std::string out = "[";
    for (std::size_t i = 0; i < sorted.size(); ++i) {
        if (i != 0) {
            out += ", ";
        }
        out += text::decimal(static_cast<uint64_t>(sorted[i]));
    }
    out += "]";
    return out;
}

/// Buses render as A / B rather than 0 / 1, matching the CSV column and the
/// other two implementations.
std::string show_buses(const std::vector<Bus>& values) {
    if (values.empty()) {
        return "none";
    }
    std::vector<Bus> sorted(values);
    std::sort(sorted.begin(), sorted.end());
    std::string out = "[";
    for (std::size_t i = 0; i < sorted.size(); ++i) {
        if (i != 0) {
            out += ", ";
        }
        out += bus_name(sorted[i]);
    }
    out += "]";
    return out;
}

void log_active_filters(const FilterConfig& filters) {
    MIE_LOG_INFO("Filtering active: exclude_types=" + show(filters.exclude_types) +
                 " exclude_rts=" + show(filters.exclude_rts) +
                 " exclude_buses=" + show_buses(filters.exclude_buses) +
                 " exclude_subaddresses=" + show(filters.exclude_subaddresses) + " include_types=" +
                 show(filters.include_types) + " include_rts=" + show(filters.include_rts) +
                 " include_buses=" + show_buses(filters.include_buses) +
                 " include_subaddresses=" + show(filters.include_subaddresses));
}

void log_filtered_out(const MieMessage& message) {
    const std::string rt = message.command_word.has_value()
                               ? text::decimal(message.command_word.value().rt)
                               : std::string("-");
    const std::string subaddress = message.command_word.has_value()
                                       ? text::decimal(message.command_word.value().subaddress)
                                       : std::string("-");
    MIE_LOG_DEBUG("Filtered out: offset=0x" + text::hex_upper(message.file_offset, 1) + " type=0x" +
                  text::hex_upper(message.type_word.message_type, 2) + " RT" + rt + " SA" +
                  subaddress + " Bus " + bus_name(message.bus()));
}

}  // namespace

bool FilterConfig::should_exclude(const MieMessage& message) const {
    const uint8_t message_type = message.type_word.message_type;
    const Bus bus = message.type_word.bus;
    const bool has_command = message.command_word.has_value();
    const uint8_t rt = has_command ? message.command_word.value().rt : 0;
    const uint8_t subaddress = has_command ? message.command_word.value().subaddress : 0;

    // Exclusion first. An RT or subaddress exclusion can only match a value the
    // record actually has, so a Command-Word-less record passes them by.
    if (contains(exclude_types, message_type)) {
        return true;
    }
    if (has_command && contains(exclude_rts, rt)) {
        return true;
    }
    if (contains(exclude_buses, bus)) {
        return true;
    }
    if (has_command && contains(exclude_subaddresses, subaddress)) {
        return true;
    }

    // Inclusion. An EMPTY set is no constraint at all -- reading it as "include
    // nothing" would make a config that sets only exclude_rts drop every record
    // in the file.
    if (!include_types.empty() && !contains(include_types, message_type)) {
        return true;
    }
    if (!include_buses.empty() && !contains(include_buses, bus)) {
        return true;
    }
    // A record with no Command Word cannot satisfy an RT or subaddress include
    // filter: an operator narrowing to RT 5 is not asking to keep records that
    // have no RT at all.
    if (!include_rts.empty() && !(has_command && contains(include_rts, rt))) {
        return true;
    }
    if (!include_subaddresses.empty() &&
        !(has_command && contains(include_subaddresses, subaddress))) {
        return true;
    }

    return false;
}

FilteredSource::FilteredSource(MessageSource& inner, const FilterConfig& filters)
    : inner_(&inner), filters_(filters), passed_(0), excluded_(0) {
    if (filters_.is_active()) {
        log_active_filters(filters_);
    }
}

FilteredSource::~FilteredSource() {
    if (!filters_.is_active()) {
        return;
    }
    // A destructor that throws during unwinding is std::terminate, and building
    // this message allocates. The tally is a courtesy; losing it must never
    // turn a decode failure into a crash.
    try {
        MIE_LOG_INFO("Filter results: " + text::decimal(passed_) + " passed, " +
                     text::decimal(excluded_) + " excluded");
    } catch (...) {  // NOLINT(bugprone-empty-catch)
    }
}

bool FilteredSource::next(MieMessage& out) {
    // Loops rather than recursing: a long excluded run must not consume stack
    // proportional to its length.
    for (;;) {
        if (!inner_->next(out)) {
            return false;
        }
        if (!filters_.should_exclude(out)) {
            passed_ += 1;
            return true;
        }
        excluded_ += 1;
        log_filtered_out(out);
    }
}

}  // namespace mie
